libvast/vast/system/partition.hpp

```diff
@@ -27,8 +32,157 @@
 
 #include <unordered_map>
 
+#include "caf/fwd.hpp"
+
 namespace vast::system {
 
+// TODO: remove this temporary namespace after we have wired all functionality
+// with the new actorized partition.
+namespace v2 {
+
+struct partition_selector {
+  bool operator()(const vast::qualified_record_field& filter,
+                  const table_slice_column& x) const;
+};
+
+/// The state of the partition actor.
+struct partition_state {
+  using partition_stream_stage_ptr = caf::stream_stage_ptr<
+    table_slice_ptr,
+    caf::broadcast_downstream_manager<
+      table_slice_column, vast::qualified_record_field, partition_selector>>;
+
+  /// Member functions
+
+  /// Gets the INDEXER at position in the layout.
+  caf::actor indexer_at(size_t position) const;
+
+  /// Retrieves an INDEXER for a predicate with a data extractor.
+  /// @param dx The extractor.
+  /// @param op The operator (only used to precompute ids for type queries.
+  /// @param x The literal side of the predicate.
+  caf::actor fetch_indexer(const data_extractor& dx, relational_operator op,
+                           const data& x) const;
+
+  /// Retrieves an INDEXER for a predicate with an attribute extractor.
+  /// @param ex The extractor.
+  /// @param op The operator (only used to precompute ids for type queries.
+  /// @param x The literal side of the predicate.
+  caf::actor fetch_indexer(const attribute_extractor& ex,
+                           relational_operator op, const data& x) const;
+
+  /// Returns all INDEXERs that are involved in evaluating the expression.
+  evaluation_triples evaluate(const expression& expr) const;
+
+  /// Data Members
+
+  /// Pointer to the parent actor.
+  caf::stateful_actor<partition_state>* self;
+
+  /// Uniquely identifies this partition.
+  uuid partition_uuid;
+
+  /// The streaming stage.
+  partition_stream_stage_ptr stage;
+
+  /// Tracks whether we already received at least one table slice.
+  bool streaming_initiated;
+
+  /// The combined type of all columns of this partition
+  record_type combined_layout;
+
+  /// Maps qualified fields to indexer actors.
+  //  TODO: Should we use the tsl map here for heterogenous key lookup?
+  detail::stable_map<qualified_record_field, caf::actor> indexers;
+
+  /// Maps type names to ids. Used the answer #type queries.
+  std::unordered_map<std::string, ids> type_ids;
+
+  /// A readable name for this partition
+  std::string name;
+
+  /// The first ID in the partition.
+  id offset;
+
+  /// The number of events in the partition.
+  size_t events;
+
+  /// Persistence-related state
+  filesystem_type fs_actor;
+  caf::response_promise persistence_promise;
+  std::optional<path> persist_path;
+  size_t persisted_indexers;
+  std::map<caf::actor_id, vast::chunk_ptr> chunks;
+};
+
+// TODO: Split this into a `static data` part that can be mmap'ed
```

Comment:
Not a huge fan of this idea tbh, for two reasons. First, semantically an `active_partition` is not an extension of a `passive_partition`, its more like a `passive_partition_builder`. (and you wouldn't let a string_builder inherit from string, etc.)

And second, the fact that they currently contain the same data is just because we didn\t get to optimizing that part yet, ideally they would have completely different memory layouts, one that is optimized for adding new data and one that is takes advantage of knowing *all* the data going into the partition and can be accessed  without deserializing.

Reply:
Good arguments, I'm convinced.